components/license-control-service/pkg/storage/storage.go

```diff
@@ -0,0 +1,261 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	_ "github.com/lib/pq" // required for pg backend
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/chef/automate/components/license-control-service/pkg/keys"
+	"github.com/chef/automate/lib/db/migrator"
+	"github.com/chef/automate/lib/logger"
+)
+
+// NoLicenseError is the error returned by a backend when it does not
+// have a configured License.
+type NoLicenseError struct{ backend string }
+
+func (n *NoLicenseError) Error() string {
+	return fmt.Sprintf("No license found in %s storage backend", n.backend)
+}
+
+// NotInitializedError is returned by a CurrentBackend when the user fails to
+// call Init() before interacting with it.
+type NotInitializedError struct{ backend string }
+
+func (n *NotInitializedError) Error() string {
+	return fmt.Sprintf("Init() never called on %s storage backend", n.backend)
+}
+
+// A CurrentBackend can be used as a storage backend for the
+// license-control-service.
+type CurrentBackend interface {
+	// Init migrates data from any known Upgradeable Backends
+	Init(context.Context, *keys.LicenseParser) error
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, NoLicense is
+	// returend.
+	GetLicense(context.Context) (string, keys.LicenseMetadata, error)
+	// SetLicense stores the given license in the backend.
+	SetLicense(context.Context, string) error
+}
+
+// An UpgradeableBackend is an old backend that can no longer be used
+// as a backend for the license-control-service but which might have
+// data that needs to be migrated into the CurrentBackend.
+type UpgradeableBackend interface {
+	// GetLicense returns the currently configured autoamte
+	// license. If no license is configured, ErrNoLicense is
+	// returend.
+	GetLicense(context.Context) (string, error)
+	// Cleanup removes the state kept by this backend. After
+	// Cleanup is called, GetLicense should return a
+	// NoLicenseError.
+	Cleanup(context.Context) error
+}
+
+var _ CurrentBackend = &PGBackend{}
+var _ UpgradeableBackend = &FileBackend{}
```

Comment:
Interesting. I'm used to `var _ UpgradeableBackend = (*FileBackened)(nil)`. Not much of a difference, though. 😄

Reply:
Oh, that is a good idea -- just to be sure that there is likely no runtime cost to this.

Reply:
Guess it's probably superfluous, when `NewCurrentBackend(...) CurrentBackend` returns `&PGBackend{...}`, it should be sufficiently checked...